blockstore/splitstore/splitstore.go

```diff
@@ -322,10 +376,34 @@ func (s *SplitStore) HashOnRead(enabled bool) {
 }
 
 func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	s.txnLk.RLock()
+	defer s.txnLk.RUnlock()
+
 	err := s.hot.View(cid, cb)
```

Comment:
OUCH, that's a very good observation; it sucks that those things aren't recursive.
We might have to write our own RWLock .

Reply:
Let me write a quick program to verify the problem.

Reply:
Yep verified the deadlock scenario; this totally sucks.

Reply:
implementation of custom txnlock that allows recursive RLocking: https://gist.github.com/vyzo/b1f6e2922270f6dfcd410a6bb06a88e5

If we actually need it, we can use it; for now let's rely on our usage patterns to not deadlock.

Reply:
Following sync discussion, the _right_ solution is to protect the references optimistically _before_ calling `s.host.View` and do the latter without holding the RLock.
This will work even if the block is missing, because protection uses `walkObjectIncomplete` which occurs checks and handles missing references.

Reply:
implemented in https://github.com/filecoin-project/lotus/pull/6474/commits/60dd97c7fc5f8f114ebebf9fe78cbb39b6672e76

Reply:
Actually it is little tricker than that, the View can theoretically be so long that it starts before a compaction starts and be active when the purge begins... which could be catastrophic.

I'll add some code to track views when a transaction is not active and block starting a new transaction while there are active views.
We can do that with a waitgroup.

Reply:
Implemented in https://github.com/filecoin-project/lotus/pull/6474/commits/acc4c374ef4226f953cfedf93121ca711f70cc4c

I think we can call the issue resolved.

Reply:
Well, add this to the pile: https://github.com/filecoin-project/lotus/pull/6474/commits/095d7427bab6063aba1f31f6cb44acb2e5ade6e9

I am leaving the conversation unresolved until we are all happy.